### PR TITLE
ENH: Min_weight for Rolling

### LINKED ISSUE #11167: ENH: `min_weight` in addition to `min_periods` for ewma
Currently the exponential functions, such as `pd.ewma`, use a `min_periods` argument to ensure there's enough data to specify generate a valid value. While this works well for the rolling functions, it's not effective for exponential functions because points have weight forever, albeit ever decreasing:

```python
In [4]: series=pd.Series(range(200))

In [5]: series[20:190]=pd.np.nan

In [6]: pd.ewma(series, span=10, min_periods=15)
Out[6]: 
0             NaN
1             NaN
2             NaN
3             NaN
4             NaN
5             NaN
6             NaN
7             NaN
8             NaN
9             NaN
10            NaN
11            NaN
12            NaN
13            NaN
14      10.277660
15      11.172348
16      12.080052
17      12.999407
18      13.929141
19      14.868084
20      14.868084
21      14.868084
22      14.868084
23      14.868084
24      14.868084
25      14.868084
26      14.868084
27      14.868084
28      14.868084
29      14.868084
          ...    
170     14.868084
171     14.868084
172     14.868084
173     14.868084
174     14.868084
175     14.868084
176     14.868084
177     14.868084
178     14.868084
179     14.868084
180     14.868084
181     14.868084
182     14.868084
183     14.868084
184     14.868084
185     14.868084
186     14.868084
187     14.868084
188     14.868084
189     14.868084
190    190.000000
191    190.550000
192    191.132890
193    191.748020
194    192.394502
195    193.071240
196    193.776953
197    194.510212
198    195.269468
199    196.053089
dtype: float64
```

I think what we want is to have a `min_weight` argument, so if you specify 0.5, it needs 50% of the weight in order to calculate a value. For rolling functions, this would be equivalent to `min_periods` being half of `window`.

What are people's thoughts?

### PR DESCRIPTION
**Pull request by max-sixty**
Initial implementation for https://github.com/pydata/pandas/issues/11167.

Needs some additional tests - for dataframes, other axis values etc, but want to get this out there and get feedback.

Note that the implemenation is different than that for `min_periods`. This uses a single function to mask all rolling calculations. `min_periods` is implemented separately for each function. I imagine the latter is a bit faster, although also more complicated.
- [x] closes #11167
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
is this in concert with ``min_periods``? (e.g. looks like this), or better as either/or?

**Comment by MaximilianR** [comment]
While you could use them together, most cases are going to be an either / or.

If it's *just* either / or, we could implement within `min_periods` - a value < 1 going to `min_weight` and otherwise going to `min_periods`. But I think less explicit.

**Comment by jreback** [comment]
``min_periods=None`` is the default, just wondering if it makes any sense to do it and/or raise if BOTH ``min_weight`` is not ``None`` and ``min_periods``. You don't need to have special values.

Furthermore does the *order* of application of ``min_periods`` and ``min_weight`` matter? 

give this some tests, if its too complicated / not worth it, ok w/o supporting both simultaneously

**Comment by MaximilianR** [comment]
> wondering if it makes any sense to do it and/or raise if BOTH min_weight is not None and min_periods.

I think it's fine to have both - it will put a `NaN` in if either test fails. They are independent, so order doesn't matter.

> give this some tests, if its too complicated / not worth it, ok w/o supporting both simultaneously

My thought exactly

File: pandas/core/window.py [hunk 1]
```diff
@@ -1373,6 +1386,19 @@ def _check_func(minp, window):
     return _check_func
 
 
+def _min_weight_mask(rolling, min_weight):
+    data = rolling.obj
```
Comment (jreback) [inline pandas/core/window.py]: can you add a comment explaining what this is doing/why

File: pandas/tests/test_window.py [hunk 2]
```diff
@@ -1398,6 +1399,58 @@ def _check_ew_structures(self, func, name):
...
+        rolling = series.expanding(min_periods=1, min_weight=0.51)
+
+        result = rolling.sum()
+        expected = Series([pd.np.NaN, pd.np.NaN, -5, pd.np.NaN, 5, 10])
+
+        assert_series_equal(result, expected)
+
+    def test_min_weight_ewm(self):
+
+        from itertools import chain
```
Comment (jreback) [inline pandas/tests/test_window.py]: import?

File: pandas/tests/test_window.py [hunk 3]
```diff
@@ -1398,6 +1399,58 @@ def _check_ew_structures(self, func, name):
         frame_result = getattr(self.frame.ewm(com=10), name)()
         self.assertEqual(type(frame_result), DataFrame)
 
+    def test_min_weight_mask_series(self):
```
Comment (jreback) [inline pandas/tests/test_window.py]: make this a separate class (you can inherit from Base)

File: pandas/tests/test_window.py [hunk 4]
```diff
@@ -1398,6 +1399,58 @@ def _check_ew_structures(self, func, name):
...
+        result = _min_weight_mask(rolling, 0.6)
+        expected = Series([False, False, True, True, True, True])
+        assert_series_equal(result, expected)
+
+        # 70%
+        result = _min_weight_mask(rolling, 0.7)
+        expected = Series([False, False, False, True, False, False])
+        assert_series_equal(result, expected)
+
+    def test_min_weight_rolling(self):
```
Comment (jreback) [inline pandas/tests/test_window.py]: add tests for _window as well

**Comment by jreback** [comment]
- update the doc-strings (Window/Rolling/Expanding/EWM), with versionadded tags.
- whatsnew sub-section to show mini-example and point to updated docs
- add to docs 
- more tests. Test for some edge cases (empty, 1 point, all-nan)

**Comment by MaximilianR** [comment]
Yes will do - this was WIP. Thanks for the comments

File: pandas/tests/test_window.py [hunk 5]
```diff
@@ -1398,6 +1399,58 @@ def _check_ew_structures(self, func, name):
...
+        from itertools import chain
+
+        # create a series with a big gap in the middle
+        series = Series(list(chain(range(9), [pd.np.NaN] * 80, range(9, 0))))
+        rolling = series.ewm(span=10, min_weight=0.5)
+
+        result = rolling.mean()
+
+        # check that all points between 25 and 90 are NaN
+        self.assertTrue(result.iloc[24:89].isnull().all())
```
Comment (jreback) [inline pandas/tests/test_window.py]: construct the actual result

**Comment by jreback** [comment]
can you rebase / update

File: pandas/core/generic.py [hunk 6]
```diff
@@ -5175,30 +5175,34 @@ def _add_series_or_dataframe_operations(cls):
         from pandas.core import window as rwindow
 
         @Appender(rwindow.rolling.__doc__)
-        def rolling(self, window, min_periods=None, freq=None, center=False,
```
Comment (jreback) [inline pandas/core/generic.py]: add to the doc-string ``min_weight`` with a versionadded tag (I don't remember exactly where these are defined)

**Comment by MaximilianR** [comment]
I've been remiss about the final fixes here. Will need a couple of weeks on other commitments but I will get to this

**Comment by jreback** [comment]
can you rebase / update?

**Comment by jreback** [comment]
if you can update

**Comment by MaximilianR** [comment]
Yes, apologies, this is another I'll attempt over / after Thanksgiving. Appreciate the patience!

**Comment by jreback** [comment]
looks like you rebased, great.

needs some docs (computation.rst), whatsnew (you can do a sub-section or just put a ref to the docs) & doc-string examples.

**Comment by jreback** [comment]
status?

**Comment by jreback** [comment]
can you update

**Comment by MaximilianR** [comment]
Yes, thanks for your patience. I'll wrap most of these up over the next month

**Comment by jreback** [comment]
@MaximilianR can you rebase / update

**Comment by MaximilianR** [comment]
Forgive the extreme delay for this + a couple others. I need a few more weeks on other stuff, and then I'll give a push to finishing these up.

**Comment by jreback** [comment]
pls rebase / update.

**Comment by jreback** [comment]
love the idea, but needs rebase and updating.
